web/src/message_store.ts

```diff
@@ -394,16 +400,204 @@ export function reify_message_id({old_id, new_id}: {old_id: number; new_id: numb
         stored_messages.set(new_id, {type: "server_message", message});
         stored_messages.delete(old_id);
+        if (message.type === "stream") {
+            remove_topic_links(message.stream_id, message.topic, message.id);
+            save_topic_links(message);
+        }
+    }
+}
+
+export type TopicLink = {
+    // Where the link occurs
+    from_message_id: number;
+    from_stream_id: number;
+    from_topic: string;
+    from_link: string;
+    // What's being linked to
+    to_stream_id: number;
+    to_topic: string | undefined;
+    to_message_id: number | undefined;
+    to_link: string;
+};
+
+const topic_links_by_from = new Map<number, Map<string, Set<TopicLink>>>();
+const topic_links_by_to = new Map<number, Map<string, Set<TopicLink>>>();
+
+function get_topic_links(
+    map: Map<number, Map<string, Set<TopicLink>>>,
+    stream_id: number,
+    topic: string,
+    create_if_missing: boolean,
+): Set<TopicLink> {
+    if (!map.has(stream_id)) {
+        if (create_if_missing) {
+            map.set(stream_id, new Map<string, Set<TopicLink>>());
+        } else {
+            return new Set();
+        }
+    }
+    if (!map.get(stream_id)!.get(topic)) {
+        if (create_if_missing) {
+            map.get(stream_id)!.set(topic, new Set<TopicLink>());
+        } else {
+            return new Set();
+        }
+    }
+    return map.get(stream_id)!.get(topic)!;
+}
+
+function set_topic_links(
+    map: Map<number, Map<string, Set<TopicLink>>>,
+    stream_id: number,
+    topic: string,
+    topic_links: Set<TopicLink>,
+): void {
+    if (!map.has(stream_id)) {
+        map.set(stream_id, new Map<string, Set<TopicLink>>());
+    }
+    map.get(stream_id)!.set(topic, topic_links);
+}
+
+export function topic_links_from_narrow(
+    stream_id: number,
+    topic: string,
+): {
+    display_html: string;
+    url: string;
+}[] {
+    const topic_links = get_topic_links(topic_links_by_from, stream_id, topic, false).values();
+    return [...topic_links].map((topic_link) =>
+        topic_link_util.get_topic_link_content_with_stream_id(
+            topic_link.to_stream_id,
+            topic_link.to_topic,
+            topic_link.to_message_id?.toString(),
+        ),
+    );
+}
+
+export function topic_links_to_narrow(
+    stream_id: number,
+    topic: string,
+): {
+    display_html: string;
+    url: string;
+}[] {
+    const topic_links = get_topic_links(topic_links_by_to, stream_id, topic, false).values();
+    return [...topic_links].map((topic_link) =>
+        topic_link_util.get_topic_link_content_with_stream_id(
+            topic_link.from_stream_id,
+            topic_link.from_topic,
+            topic_link.from_message_id.toString(),
+        ),
+    );
+}
+
+function filter_set<T>(set: Set<T>, filter: (item: T) => boolean): Set<T> {
+    return new Set([...set].filter((item) => filter(item)));
+}
+
+function remove_topic_links(stream_id: number, topic: string, message_id: number): void {
```

Comment:
I feel like this `remove_topic_links` and `save_topic_links` should maybe be async? I haven't done performance analysis on it yet and don't know how high priority the message edit/send is for being fast (especially since we have the local echo already)

Reply:
If the data structures are done right, it should be O(1) to find the records to adjust. We should just avoid iterating through huge data structures.

Reply:
`remove_topic_links` is currently O(# of messages with links in the topic), which isn't ideal. But I initially tried storing them by message id and remembered not all links have an associated message id (e.g. links to topics and links to channels). But I just added back the map for storing by message id. I used `NO_MESSAGE_ID = 0` for when the `message_id` isn't set, does that seem reasonable?

Reply:
@timabbott does this feel resolved?